lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp

```diff
@@ -471,9 +480,37 @@ class Inliner {
 
   /// A mapping of module names to NLA symbols that originate from that module.
   DenseMap<Attribute, SmallVector<Attribute>> rootMap;
+
+  /// The current instance path.  This is a pair<ModuleName, InstanceName>.
+  /// This is used to distinguish if a non-local annotation applies to the
+  /// current instance or not.
+  SmallVector<std::pair<Attribute, Attribute>> currentPath;
 };
 } // namespace
 
+/// Check if the NLA applies to our instance path. This works by verifying the
+/// instance paths backwords starting from the current module. We drop the back
+/// element from the NLA because it obviously matches the current operation.
+bool Inliner::doesNLAMatchCurrentPath(NonLocalAnchor nla) {
+  auto nlaPath = nla.namepath().getValue().drop_back();
+  auto nlaIt = nlaPath.rbegin();
+  auto nlaEnd = nlaPath.rend();
+  auto pathIt = currentPath.rbegin();
+  auto pathEnd = currentPath.rend();
+  while (nlaIt != nlaEnd && pathIt != pathEnd) {
+    auto innerRef = (*nlaIt++).cast<hw::InnerRefAttr>();
+    auto &pair = (*pathIt++);
```

Comment:
nit: `auto &[module, name] = *pathIt++` so we know what `first` and `second` are